✨ Ignition: add option to store User Data in plain text

damdo opened the pull request:
<!-- Thanks for this PR! If this is your first PR please read the [contributing guide](../CONTRIBUTING.md) -->
<!-- If this PR is still work-in-progress and is being open for visibility please prefix the title with `WIP:` -->

**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:

For Machine bootstrapping CAPA supports one main method: CloudInit and one experimental method: Ignition.

For the former, CAPA used to store the User Data that the bootstrapper would use to fulfill its job in the EC2 Instance User Data in plain text.

This was not considered the most secure way to provide this configuration to a machine as a process/subject with the power to query the instance metadata could be able to read this config, which may contain certs, keys and other forms of secrets.

As such to minimize this issue, a new default method was added, where the controller would now leverage AWS SSM to encrypt the user data before storing it in the EC2 Instance User Data. The previous method of storing User Data in plain text was kept available as an alternative option to users, which would be able to access it by providing the `InsecureSkipSecretsManager` option in the AWSMachine spec.

Later on, when the new Ignition bootstrapping support was added, the creators, already aware of the security implications of storing User Data in plain text,
decided to opt for a more secure default. After briefly looking at using AWS SSM encryption also for Ignition, they acknowledged that it wasn't possible due to Ignition's lack of support for multi mime types. Hence they went for using S3 Buckets to store the config in objects, and have ignition fetch it at bootstrap time from there, and this did the trick.

Up until now this has been the only option to achieve Ignition bootstrapping.
However even if this is a better option from a security standpoint than storing User Data in plain text, and as such the preferred option for a good default,
it requires the ability to provision/access S3 Buckets, which is not always possible/desirable in specific scenarios, as well as not suitable for certain type of bespoke CAPI bootstrap providers.

As such we would like to introduce for Ignition, similarly to what is available for CloudInit, the ability to store User Data in the EC2 Instance in plain text.
Of course we know this is not the best option from a security standpoint, and as such we want to make it a fallback option that users can leverage if their use case does not allow the use of S3 buckets. We mentioned in the docs, and in the API that this option is discouraged and we trust CAPA users will use this with care.



**Special notes for your reviewer**:

**Checklist**:
<!-- Put an "X" character inside the brackets of each completed task. Some may be optional depending on the PR in which case these can be deleted -->

- [x] squashed commits
- [x] includes documentation
- [x] includes [emojis](https://github.com/kubernetes-sigs/kubebuilder-release-tools?tab=readme-ov-file#kubebuilder-project-versioning)
- [x] adds unit tests

**Release note**:
```release-note
Ignition: add option to store User Data in plain text
```

damdo commented:
/cc @JoelSpeed @vincepri @nrb

damdo commented:
/test pull-cluster-api-provider-aws-test

dlipovetsky commented:
If the user is selecting a storage mechanism, then I think the two choices are:
- S3
- Instance Userdata

https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/instancedata-add-user-data.html

So I think `InstanceUserdata` would be a good name.

richardcase commented:
/milestone v2.4.0

Ankitasw commented on docs/book/src/topics/ignition-support.md:
```diff
@@ -23,17 +23,57 @@ For more generic information, see [Cluster API documentation on Ignition Bootstr
+When using CloudInit for bootstrapping, by default the awsmachine controller stores EC2 instance user data using SSM to store it encrypted, which underneath uses multi part mime types.
+Unfortunately multi part mime types are [not supported](https://github.com/coreos/ignition/issues/1072) by Ignition. Moreover EC2 instance user data storage is also limited to 64 KB, which might not always be enough to provision Kubernetes controlplane because of the size of required certificates and configuration files.
 
-EC2 user data is also limited to 64 KB, which is often not enough to provision Kubernetes controlplane because
```
This statement still holds good, isn't it?

damdo replied:
It does and in fact it is still there I moved it in the paragraph above 🙂

Ankitasw commented on docs/book/src/topics/ignition-support.md:
```diff
@@ -51,22 +91,7 @@ spec:
 See [Using clusterawsadm to fulfill prerequisites](./using-clusterawsadm-to-fulfill-prerequisites.md) for more
 details.
 
-## Enabling EXP_BOOTSTRAP_FORMAT_IGNITION feature gate
```
I might have missed, but is the feature gate removed for ignition support?

damdo replied:
I haven't removed it 👍 it is just a bit higher up in the doc now

Ankitasw commented on docs/book/src/topics/ignition-support.md:
```diff
@@ -109,6 +134,27 @@ spec:
 
+### Store Ignition config as PlainTextUserData
+
+WARNING: this is discouraged as is not considered as secure as other storage types.
```
Would appreciate if this warning is highlighted more, like in bold/code blocked.

damdo replied:
Added it in a warning note block and set it as bold text.

Ankitasw reviewed (commented):
This suggestion seems ok to me if we have a valid use case around this. 
Overall code is also good, some small nits/questions from my end.

damdo commented:
@dlipovetsky @richardcase as discussed during the CAPA meeting I changed the storageType value from `PlainTextUserData` to `InstanceUserData` (cc. @JoelSpeed)

@Ankitasw I've addressed your comments 👍 

Thanks all for the input here.

vincepri commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -197,6 +208,26 @@ type Ignition struct {
+	// that can access the ec2 metadata service have access to this sensitive information.
+	// So this is only to be used at ones own risk, and only when other more secure options are not viable.
+	//
+	// +optional
+	// +kubebuilder:default="S3"
+	// +kubebuilder:validation:Enum:="S3";"InstanceUserData"
+	StorageType IgnitionStorageTypeOption `json:"storageType,omitempty"`
```
```suggestion
	// +optional
	UseUnencryptedUserData bool `json:"useUnencryptedUserData,omitempty"`
```

vincepri commented on controllers/awsmachine_controller.go:
```diff
@@ -745,7 +754,9 @@ func (r *AWSMachineReconciler) cloudInitUserData(machineScope *scope.MachineScop
 }
 
-func (r *AWSMachineReconciler) ignitionUserData(scope *scope.MachineScope, objectStoreSvc services.ObjectStoreInterface, userData []byte) ([]byte, error) {
+// ignitionUserDataWithS3 creates an S3 bucket object and stores user data in it,
+// then returns the config to instruct ignition how to pull the user data from S3.
+func (r *AWSMachineReconciler) ignitionUserDataWithS3(scope *scope.MachineScope, objectStoreSvc services.ObjectStoreInterface, userData []byte) ([]byte, error) {
```
```suggestion
// generateIgnitionConfigRemoteStorage uses a remote storage (S3 bucket) object and stores user data in it,
// then returns the config to instruct ignition how to pull the user data from the bucket.
func (r *AWSMachineReconciler) generateIgnitionConfigRemoteStorage(scope *scope.MachineScope, objectStoreSvc services.ObjectStoreInterface, userData []byte) ([]byte, error) {
```

vincepri commented on controllers/awsmachine_controller.go:
```diff
@@ -745,7 +754,9 @@ func (r *AWSMachineReconciler) cloudInitUserData(machineScope *scope.MachineScop
+// then returns the config to instruct ignition how to pull the user data from S3.
+func (r *AWSMachineReconciler) ignitionUserDataWithS3(scope *scope.MachineScope, objectStoreSvc services.ObjectStoreInterface, userData []byte) ([]byte, error) {
 	if objectStoreSvc == nil {
 		return nil, errors.New("object store service not available")
```
```suggestion
		return nil, errors.New("using Ignition requires an object storage by default, this can be se in `AWSCluster.Spec.Ignition.S3Bucket`, or to use EC2 user-data set `AWSMachine.Spec. .Ignition.UseUnencryptedUserData`")
```

vincepri commented on api/v1beta2/awsmachine_types.go:
```diff
@@ -197,6 +208,26 @@ type Ignition struct {
+	// that can access the ec2 metadata service have access to this sensitive information.
+	// So this is only to be used at ones own risk, and only when other more secure options are not viable.
+	//
+	// +optional
+	// +kubebuilder:default="S3"
+	// +kubebuilder:validation:Enum:="S3";"InstanceUserData"
+	StorageType IgnitionStorageTypeOption `json:"storageType,omitempty"`
```
```suggestion
	// +optional
	// +kubebuilder:default="ClusterObjectStore"
	// +kubebuilder:validation:Enum:="ClusterObjectStore";"UnencryptedUserData"
	StorageType IgnitionStorageTypeOption `json:"storageType,omitempty"`
```

Proposing that we do away, at this point, from choosing `S3` as an option here because the object storage is defined on the AWSCluster object.

I'd also propose to rename `InstanceUserData` to `UnencryptedUserData` instead, given that the current implementation can be expanded in the future if encryption options are available for Ignition, and `UserData` is a well-known concept associated with an EC2 Instance already.

@damdo @dlipovetsky @richardcase thoughts?

damdo replied:
Hey @vincepri , thanks for reviewing this.
As we discussed offline, I'm happy with changing the enum value `S3` to `ClusterObjectStore`, as that can give more insight into the fact that the Object store details are defined at a Cluster level on the Cluster object.

As for the `UnencryptedUserData` vs `InstanceUserData`, I don't have strong opinions, can go either way. So I'm happy to change provided others are too.

Let's see if @dlipovetsky and @richardcase agree.

richardcase replied:
I think we need to think consider whether changing the enum here is considered backwards compatible. Especially as this is now potentially suggesting:

- Adding a new enum value (i.e. `UnencryptedUserData`)
- Changing an existing one. Is `ClusterObjectStore` schematically equivalent to `S3`?
- Changing the default value

Let's review the [Changing the API](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md) doc. If we don't have a compatible change, then we'd need to bump the API version.

Its been a whilst since i read the doc, so i will do that ASAP.

damdo replied:
@richardcase API wise, the `storageType` Enum is a new field that I'm adding under the existing `ignition` field, so in that regard it is an addition (no breaking change).

Behaviour wise, up until now, there was no way to change the default behaviour, which was to use the Cluster Object Storage reference defined under the `AWSCluster` object spec [here](https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/f8b9cfd43ff0dd3e2887c3acad3d1ac53284cf8c/api/v1beta2/awscluster_types.go#L98-L103).
From now on, with this new Enum we will provide a way to tweak this.
If this is not set (default) or `ClusterObjectStore` is set, the user will get the current behaviour.
If `UnencryptedUserData` is set, the user will get a different behaviour (storing of the value unencrypted directly in the AWS User Data).

So no breaking change here both from an API and from a behaviour standpoints, default remains the same, only an extra option for different behaviour.

richardcase replied:
Great, thanks for explaining that @damdo 👍 I will take time to properly review asap.

damdo replied:
Thanks @richardcase
I haven't yet included the above Enum values renaming that @vincepri is proposing, but I intend to, if we are all happy with it.
So once you let me know, if you are happy, I can change this as well before the final review.

damdo replied:
After chatting with @nrb @vincepri @richardcase and @dlipovetsky today at the CAPA meeting we decided to go with the Enum values suggested by @vincepri. I'll update the PR tomorrow with those values.

richardcase commented:
/milestone v2.4.0

vincepri reviewed (approved):
/lgtm

Ankitasw reviewed (commented):
@damdo do you think we can somehow test this in any of the existing E2E test case? I think right now for ignition we have only one E2E tests with S3.

damdo commented:
@Ankitasw Yeah I had a look at this. We are testing ignition in combination with creating the S3 bucket and bootstrapping from it. Although I wasn't sure of the value of testing if plainly setting the user data field in the AWS Instance SDK and checking if that works. This behaviour is basically a subset of the Ignition + S3 (where we do ignition with a basic user data config stored in the instance user data + a boot hook S3), so I considered it covered by the other case.
That said if we feel like we want this I can have a look.

vincepri commented:
We can add a small additional case to the current e2e where the s3 bucket isn't used, but instead we use user data? Create a single machine

damdo commented:
/test pull-cluster-api-provider-aws-e2e

nrb commented on test/e2e/suites/unmanaged/unmanaged_functional_test.go:
```diff
@@ -1070,6 +1071,42 @@ var _ = ginkgo.Context("[unmanaged] [functional]", func() {
+			Expect(len(unencryptedUDWorkerMachines)).To(Equal(1))
+			// There is only one machine.
+			m := unencryptedUDWorkerMachines[0]
+			mScope := &scope.MachineScope{Machine: &m}
```
From CI:

```
./unmanaged_functional_test.go:1104:72: cannot use m (variable of type "sigs.k8s.io/cluster-api/api/v1beta1".Machine) as *"sigs.k8s.io/cluster-api/api/v1beta1".Machine value in struct literal
./unmanaged_functional_test.go:1104:75: cannot call pointer method GetRawBootstrapDataWithFormat on "sigs.k8s.io/cluster-api-provider-aws/v2/pkg/cloud/scope".MachineScope
./unmanaged_functional_test.go:1107:38: cannot use m.Spec.ProviderID (variable of type *string) as string value in argument to assertUnencryptedUserDataIgnition
```

damdo replied:
Yes 👍  I'm fixing those.
Unfortunately my editor's language server seems to be broken at the moment, and it is not catching/reporting these issues, so CI is my LSP atm 😓

damdo replied:
Fixed. Thanks

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

richardcase commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

nrb commented:
465 instances of `failed to add instance to Event Bridge instance state rule`

```
E0205 18:23:52.359879       1 controller.go:329] "Reconciler error" err="failed to add instance to Event Bridge instance state rule: unable to describe rule functional-test-ssm-parameter-store-75xzqz-ec2-rule: ResourceNotFoundException: Rule functional-test-ssm-parameter-store-75xzqz-ec2-rule does not exist on EventBus default." controller="awsmachine" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSMachine" AWSMachine="functional-test-ssm-parameter-store-dheeri/functional-test-ssm-parameter-store-75xzqz-control-plane-w74sg" namespace="functional-test-ssm-parameter-store-dheeri" name="functional-test-ssm-parameter-store-75xzqz-control-plane-w74sg" reconcileID="c633a078-8f9c-42b3-ab09-9840a95cacc1"
```

Also 160 references to `panic`, starting at 18:19:18 and ending at 19:05:43.

```
E0205 18:19:18.143014       1 runtime.go:79] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 609 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic({0x2e79c20?, 0x5a66710})
	/go/pkg/mod/k8s.io/apimachinery@v0.28.4/pkg/util/runtime/runtime.go:75 +0x85
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:108 +0xb2
panic({0x2e79c20?, 0x5a66710?})
	/usr/local/go/src/runtime/panic.go:914 +0x21f
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).resolveUserData(0xc0005c8120, 0xc0017c13e0, {0x3bdd638, 0xc001fdde00}, {0x3bbe520, 0xc0015be0f0})
	/workspace/controllers/awsmachine_controller.go:706 +0x134
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).createInstance(0xc0017c13e0?, {0x3bdd718, 0xc000aa4b40}, 0xc0017c13e0, {0x3bdd638, 0xc001fdde00}, {0x3bbe520, 0xc0015be0f0})
	/workspace/controllers/awsmachine_controller.go:681 +0x86
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).reconcileNormal(0xc0005c8120, {0x3bea538?, 0xc001fdde00?}, 0xc0017c13e0, {0x3bdd638, 0xc001fdde00}, {0x3bea538, 0xc001fdde00}, {0x3be92b0, 0xc001fdde00}, ...)
	/workspace/controllers/awsmachine_controller.go:517 +0x338
sigs.k8s.io/cluster-api-provider-aws/v2/controllers.(*AWSMachineReconciler).Reconcile(0xc0005c8120, {0x3bbe1f8, 0xc00226f8c0}, {{{0xc002408600?, 0x0?}, {0xc001a56100?, 0xc0020c3d48?}}})
	/workspace/controllers/awsmachine_controller.go:235 +0xac9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x3bc2560?, {0x3bbe1f8?, 0xc00226f8c0?}, {{{0xc002408600?, 0xb?}, {0xc001a56100?, 0x0?}}})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:119 +0xb7
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0005b3ae0, {0x3bbe230, 0xc00080b5e0}, {0x3025480?, 0xc001963ec0?})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:316 +0x3cc
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0005b3ae0, {0x3bbe230, 0xc00080b5e0})
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:266 +0x1c9
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:227 +0x79
created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2 in goroutine 52
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.16.3/pkg/internal/controller/controller.go:223 +0x565
E0205 18:19:18.143659       1 controller.go:329] "Reconciler error" err="panic: runtime error: invalid memory address or nil pointer dereference [recovered]" controller="awsmachine" 
```

https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-aws/4700/pull-cluster-api-provider-aws-e2e/1754565031283920896/artifacts/clusters/bootstrap/controllers/capa-controller-manager/capa-controller-manager-6b8f8b488c-8fzxd/manager.log

damdo commented:
@nrb Thanks for digging. I'm trying to isolate the e2e issue I introduced from the flakes. Let's see if a rerun helps.


/test pull-cluster-api-provider-aws-e2e

vincepri commented on controllers/awsmachine_controller.go:
```diff
@@ -701,7 +701,16 @@ func (r *AWSMachineReconciler) resolveUserData(machineScope *scope.MachineScope,
 
 	if machineScope.UseIgnition(userDataFormat) {
-		userData, err = r.ignitionUserData(machineScope, objectStoreSvc, userData)
+		ignitionStorageType := machineScope.AWSMachine.Spec.Ignition.StorageType
```
Seems this could still produce a nil pointer exception? The above is an OR between the secret type and spec.ignition

damdo commented:
/test pull-cluster-api-provider-aws-test

damdo commented:
/test pull-cluster-api-provider-aws-e2e

richardcase commented:
I'm seeing this in the CAPA logs for the failing e2e:

```
E0206 11:48:06.531883       1 awsmachine_controller.go:519] "unable to create instance" err="failed to resolve userdata: unsupported ignition storageType \"\""
E0206 11:48:06.532561       1 controller.go:329] "Reconciler error" err="failed to resolve userdata: unsupported ignition storageType \"\"" controller="awsmachine" controllerGroup="infrastructure.cluster.x-k8s.io" controllerKind="AWSMachine" AWSMachine="functional-test-ignition-w6200e/functional-test-ignition-hqv6ar-control-plane-bqqxz" namespace="functional-test-ignition-w6200e" name="functional-test-ignition-hqv6ar-control-plane-bqqxz" reconcileID="5d081c07-ba4e-40c1-8eb7-b4d3aa441a37"
```

damdo commented:
/test pull-cluster-api-provider-aws-e2e

nrb commented:
/test pull-cluster-api-provider-aws-e2e
time out

vincepri commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
Found the issue with the deprovisioning, let's see if E2Es pass now:

/test pull-cluster-api-provider-aws-e2e

damdo commented:
@vincepri @nrb @richardcase "old" e2e passed 🎉 
I'll re-add the new e2e case for unencrypted UserData, and see if that works.

damdo commented:
/test pull-cluster-api-provider-aws-e2e

vincepri commented on test/e2e/suites/unmanaged/unmanaged_functional_test.go:
```diff
@@ -1070,6 +1071,42 @@ var _ = ginkgo.Context("[unmanaged] [functional]", func() {
+			// There is only one machine.
+			m := unencryptedUDWorkerMachines[0]
+			mScope := &scope.MachineScope{Machine: &m}
+			machineUserData, userDataFormat, err := mScope.GetRawBootstrapDataWithFormat()
```
This is returning a nil pointer exception in the test

vincepri replied:
```
runtime error: invalid memory address or nil pointer dereference

Full Stack Trace
  sigs.k8s.io/cluster-api-provider-aws/v2/pkg/cloud/scope.(*MachineScope).GetRawBootstrapDataWithFormat(0xc0008578e0)
  	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/pkg/cloud/scope/machine.go:276 +0x6f
  sigs.k8s.io/cluster-api-provider-aws/v2/test/e2e/suites/unmanaged.glob..func6.15.1()
  	/home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/suites/unmanaged/unmanaged_functional_test.go:1105 +0xd68
```

damdo replied:
Hm yeah fixed it below.

vincepri commented on test/e2e/suites/unmanaged/unmanaged_functional_test.go:
```diff
@@ -1070,6 +1071,42 @@ var _ = ginkgo.Context("[unmanaged] [functional]", func() {
+			Expect(len(unencryptedUDWorkerMachines)).To(Equal(1))
+			// There is only one machine.
+			m := unencryptedUDWorkerMachines[0]
+			mScope := &scope.MachineScope{Machine: &m}
```
The `MachineScope` seems it needs a client, but it's a private field; could we achieve the same in a different way?

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

damdo commented:
/test pull-cluster-api-provider-aws-e2e

vincepri commented:
Seems another test is failing:

```
STEP: Waiting for the workload nodes to exist - /home/prow/go/pkg/mod/sigs.k8s.io/cluster-api/test@v1.6.1/framework/machinedeployment_helpers.go:106 @ 02/08/24 15:22:47.926
[FAILED] Expected
    <string>: cloud-config
to equal
    <string>: ignition
In [It] at: /home/prow/go/src/sigs.k8s.io/cluster-api-provider-aws/test/e2e/suites/unmanaged/unmanaged_functional_test.go:1105 @ 02/08/24 15:23:48.067
```

damdo commented:
> Seems another test is failing:

@vincepri 
Thanks, yes I noticed that previously and fixed it 👍  We should be quite close to succeeding now 🤞 
Let's see what the pending e2e run reports.

vincepri reviewed (approved):
/lgtm

Great work!

/assign @Ankitasw @richardcase 
for approval

Ankitasw commented:
/approve

Thankyou @damdo for your work 👏
